app/uk/gov/hmrc/exports/services/reversemapping/declaration/locations/GoodsLocationParser.scala

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.services.reversemapping.declaration.locations
+
+import uk.gov.hmrc.exports.models.declaration.GoodsLocation
+import uk.gov.hmrc.exports.services.reversemapping.MappingContext
+import uk.gov.hmrc.exports.services.reversemapping.declaration.DeclarationXmlParser
+import uk.gov.hmrc.exports.services.reversemapping.declaration.DeclarationXmlParser._
+import uk.gov.hmrc.exports.services.reversemapping.declaration.XmlTags.{GoodsLocation => GoodsLocationTag, _}
+
+import javax.inject.Singleton
+import scala.xml.NodeSeq
+
+@Singleton
+class GoodsLocationParser extends DeclarationXmlParser[Option[GoodsLocation]] {
+
+  override def parse(inputXml: NodeSeq)(implicit context: MappingContext): XmlParserResult[Option[GoodsLocation]] =
+    Right((inputXml \ Declaration \ GoodsShipment \ Consignment \ GoodsLocationTag).headOption.flatMap(parseGoodsLocation))
+
+  private def parseGoodsLocation(goodsLocationElement: NodeSeq): Option[GoodsLocation] = {
+    val country = (goodsLocationElement \ Address \ CountryCode).toStringOption
+    val typeOfLocation = (goodsLocationElement \ TypeCode).toStringOption
+    val qualifierOfIdentification = (goodsLocationElement \ Address \ TypeCode).toStringOption
+    val identificationOfLocation = (goodsLocationElement \ Name).toStringOption
+
+    if (Seq(country, typeOfLocation, qualifierOfIdentification).flatten.isEmpty)
```

Comment:
Out of curiosity... What if `identificationOfLocation` is `nonEmpty`?

Reply:
Didn't think about it... I'll add a test case for that and I'll change it to return incomplete `GoodsLocation` object. This is more in line with what is already implemented.